Do doctests test suites always need to depend on the library?

It seems that `doctests` test suites might always have an implicit dependency on the library being tested. I noticed this when doing the following:

```
$ cabal get parsers-0.12.5
Unpacking to parsers-0.12.5/
$ cd parsers-0.12.5/
$ cabal configure --enable-tests
Resolving dependencies...
[1 of 1] Compiling Main             ( dist/setup/setup.hs, dist/setup/Main.o )
Linking ./dist/setup/setup ...
Configuring parsers-0.12.5...
$ cabal test doctests
Preprocessing test suite 'doctests' for parsers-0.12.5..
Building test suite 'doctests' for parsers-0.12.5..

<no location info>: warning: [-Wmissing-home-modules]
    These modules are needed for compilation but not listed in your .cabal file's other-modules: Build_doctests
[1 of 2] Compiling Build_doctests   ( dist/build/doctests/autogen/Build_doctests.hs, dist/build/doctests/doctests-tmp/Build_doctests.o )
[2 of 2] Compiling Main             ( tests/doctests.hs, dist/build/doctests/doctests-tmp/Main.o )

<no location info>: warning: [-Wmissing-home-modules]
    These modules are needed for compilation but not listed in your .cabal file's other-modules: Build_doctests
Linking dist/build/doctests/doctests ...
Running 1 test suites...
Test suite doctests: RUNNING...

<command-line>:7:0: error:
     fatal error: dist/build/autogen/cabal_macros.h: No such file or directory
compilation terminated.
-i
-i/home/rgscott/Documents/Hacking/Haskell/parsers-0.12.5/dist/build/autogen
-i/home/rgscott/Documents/Hacking/Haskell/parsers-0.12.5/dist/build
-i/home/rgscott/Documents/Hacking/Haskell/parsers-0.12.5/src
-hide-all-packages
-package-db=dist/package.conf.inplace
-optP-include
-optPdist/build/autogen/cabal_macros.h
-package-id=base-4.10.0.0
-package-id=base-orphans-0.6-9iPDLrUqQ5yHb9fb9OUZ9L
-package-id=charset-0.3.7.1-GkbAILH2Vj9HXMMGXPSeNg
-package-id=containers-0.5.10.2
-package-id=parsec-3.1.11-DPgnR92AWEaFOaixmwipet
-package-id=attoparsec-0.13.1.0-2xtipSJ3oFDBfLwedqccX
-package-id=text-1.2.2.2-EGUst8sqNAZCw1xLPcmcMH
-package-id=transformers-0.5.2.0
-package-id=mtl-2.2.1-19EL8AGBsN3DnnOhrC9xY3
-package-id=scientific-0.3.5.1-L0qYzdp0wpz8rO8gircNSR
-package-id=unordered-containers-0.2.8.0-HVcKYx0GQoVIIbKuoCcJbx
-package-id=bytestring-0.10.8.2
-package-id=directory-1.3.0.2
-package-id=doctest-0.12.0-JcuLvnckae744xvqjXdMWF
-package-id=filepath-1.4.1.2
-package-id=QuickCheck-2.10.0.1-DTIBC3CyU6p3h4xSnOjkg1
-package-id=quickcheck-instances-0.3.16-A15GUudjPiB6fV2AyfxwM5
Text.Parser.Char
Text.Parser.Combinators
Text.Parser.LookAhead
Text.Parser.Permutation
Text.Parser.Expression
Text.Parser.Token
Text.Parser.Token.Style
Text.Parser.Token.Highlight
doctests: `gcc' failed in phase `C pre-processor'. (Exit code: 1)
Test suite doctests: FAIL
Test suite logged to: dist/test/parsers-0.12.5-doctests.log
0 of 1 test suites (0 of 1 test cases) passed.
```

The culprit appears to be that the `doctests` test suite in `parsers` lacks an explicit dependency on the `parsers` library. If I add that, then `cabal test doctests` works as expected.

Currently, we advise users in the documentation to depend on the library in a `doctests` test suite only in the event that one uses `cbits`, but perhaps we should always be advising this?

> @RyanGlScott can you paste the output of `cabal test doctest` (the test.log) in successful case?

Here's the full output of `cabal test doctests`:

```
$ cabal test doctests
[1 of 1] Compiling Main             ( dist/setup/setup.hs, dist/setup/Main.o ) [flags changed]
Linking ./dist/setup/setup ...
Preprocessing library for parsers-0.12.5..
Building library for parsers-0.12.5..
[1 of 8] Compiling Text.Parser.Combinators ( src/Text/Parser/Combinators.hs, dist/build/Text/Parser/Combinators.o )
[2 of 8] Compiling Text.Parser.Char ( src/Text/Parser/Char.hs, dist/build/Text/Parser/Char.o )
[3 of 8] Compiling Text.Parser.Expression ( src/Text/Parser/Expression.hs, dist/build/Text/Parser/Expression.o )

src/Text/Parser/Expression.hs:115:15: warning: [-Wmonomorphism-restriction]
    • The Monomorphism Restriction applies to the bindings for ‘rassoc’,
                                                               ‘lassoc’, ‘nassoc’, ‘prefix’,
                                                               ‘postfix’
        Consider giving a type signature for these binders
    • In the expression:
        let
          rassoc, lassoc, nassoc :: [m (a -> a -> a)]
          prefix, postfix :: [m (a -> a)]
          (rassoc, lassoc, nassoc, prefix, postfix)
            = foldr splitOp ([], [], [], [], []) ops
          ....
        in
          termP <**> (rassocP <|> lassocP <|> nassocP <|> pure id)
            <?> "operator"
      In an equation for ‘makeParser’:
          makeParser term ops
            = let
                rassoc, lassoc, nassoc :: [m (a -> a -> a)]
                prefix, postfix :: [m (a -> a)]
                ....
              in
                termP <**> (rassocP <|> lassocP <|> nassocP <|> pure id)
                  <?> "operator"
      In an equation for ‘buildExpressionParser’:
          buildExpressionParser operators simpleExpr
            = foldl makeParser simpleExpr operators
            where
                makeParser term ops
                  = let ...
                    in
                      termP <**> (rassocP <|> lassocP <|> nassocP <|> pure id)
                        <?> "operator"
                splitOp (Infix op assoc) (rassoc, lassoc, nassoc, prefix, postfix)
                  = case assoc of
                      AssocNone -> ...
                      AssocLeft -> ...
                      AssocRight -> ...
                splitOp (Prefix op) (rassoc, lassoc, nassoc, prefix, postfix)
                  = (rassoc, lassoc, nassoc, op : prefix, postfix)
                splitOp (Postfix op) (rassoc, lassoc, nassoc, prefix, postfix)
                  = (rassoc, lassoc, nassoc, prefix, op : postfix)
    |
115 |               (rassoc,lassoc,nassoc,prefix,postfix) = foldr splitOp ([],[],[],[],[]) ops
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[4 of 8] Compiling Text.Parser.LookAhead ( src/Text/Parser/LookAhead.hs, dist/build/Text/Parser/LookAhead.o )
[5 of 8] Compiling Text.Parser.Permutation ( src/Text/Parser/Permutation.hs, dist/build/Text/Parser/Permutation.o )
[6 of 8] Compiling Text.Parser.Token.Highlight ( src/Text/Parser/Token/Highlight.hs, dist/build/Text/Parser/Token/Highlight.o )
[7 of 8] Compiling Text.Parser.Token ( src/Text/Parser/Token.hs, dist/build/Text/Parser/Token.o )

src/Text/Parser/Token.hs:642:38: warning: [-Wincomplete-uni-patterns]
    Pattern match(es) are non-exhaustive
    In a pattern binding: Patterns not matched: (_, [])
    |
642 |                                      (low, ex : high) = splitAt bd dps
    |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[8 of 8] Compiling Text.Parser.Token.Style ( src/Text/Parser/Token/Style.hs, dist/build/Text/Parser/Token/Style.o )
Preprocessing test suite 'doctests' for parsers-0.12.5..
Building test suite 'doctests' for parsers-0.12.5..

<no location info>: warning: [-Wmissing-home-modules]
    These modules are needed for compilation but not listed in your .cabal file's other-modules: Build_doctests
[1 of 2] Compiling Build_doctests   ( dist/build/doctests/autogen/Build_doctests.hs, dist/build/doctests/doctests-tmp/Build_doctests.o )
[2 of 2] Compiling Main             ( tests/doctests.hs, dist/build/doctests/doctests-tmp/Main.o )

<no location info>: warning: [-Wmissing-home-modules]
    These modules are needed for compilation but not listed in your .cabal file's other-modules: Build_doctests
Linking dist/build/doctests/doctests ...
Running 1 test suites...
Test suite doctests: RUNNING...
Test suite doctests: PASS
Test suite logged to: dist/test/parsers-0.12.5-doctests.log
1 of 1 test suites (1 of 1 test cases) passed.
```

And here's the contents of `dist/test/parsers-0.12.5-doctests.log `:

```
$ more dist/test/parsers-0.12.5-doctests.log 
Test suite doctests: RUNNING...                                                           
Examples: 0  Tried: 0  Errors: 0  Failures: 0                                             
-i                                                                                        
-i/home/rgscott/Documents/Hacking/Haskell/parsers-0.12.5/dist/build/autogen               
-i/home/rgscott/Documents/Hacking/Haskell/parsers-0.12.5/dist/build                       
-i/home/rgscott/Documents/Hacking/Haskell/parsers-0.12.5/src                              
-hide-all-packages                                                                        
-package-db=dist/package.conf.inplace                                                     
-optP-include                                                                             
-optPdist/build/autogen/cabal_macros.h                                                    
-package-id=base-4.10.0.0                                                                 
-package-id=base-orphans-0.6-9iPDLrUqQ5yHb9fb9OUZ9L                                       
-package-id=charset-0.3.7.1-GkbAILH2Vj9HXMMGXPSeNg                                        
-package-id=containers-0.5.10.2                                                           
-package-id=parsec-3.1.11-DPgnR92AWEaFOaixmwipet                                          
-package-id=attoparsec-0.13.1.0-2xtipSJ3oFDBfLwedqccX                                     
-package-id=text-1.2.2.2-EGUst8sqNAZCw1xLPcmcMH                                           
-package-id=transformers-0.5.2.0                                                          
-package-id=mtl-2.2.1-19EL8AGBsN3DnnOhrC9xY3
-package-id=scientific-0.3.5.1-L0qYzdp0wpz8rO8gircNSR
-package-id=unordered-containers-0.2.8.0-HVcKYx0GQoVIIbKuoCcJbx
-package-id=bytestring-0.10.8.2
-package-id=directory-1.3.0.2
-package-id=doctest-0.12.0-JcuLvnckae744xvqjXdMWF
-package-id=filepath-1.4.1.2
-package=parsers-0.12.5
-package-id=QuickCheck-2.10.0.1-DTIBC3CyU6p3h4xSnOjkg1
-package-id=quickcheck-instances-0.3.16-A15GUudjPiB6fV2AyfxwM5
Text.Parser.Char
Text.Parser.Combinators
Text.Parser.LookAhead
Text.Parser.Permutation
Text.Parser.Expression
Text.Parser.Token
Text.Parser.Token.Style
Text.Parser.Token.Highlight
Test suite doctests: PASS
Test suite logged to: dist/test/parsers-0.12.5-doctests.log
```

> And also `find dist` ?

```
$ find dist
dist                                                                                      
dist/package.conf.inplace                                                                 
dist/package.conf.inplace/parsers-0.12.5-BbWYl7dZaXs10a39DfSMvp.conf                      
dist/package.conf.inplace/package.cache.lock                                              
dist/package.conf.inplace/package.cache                                                   
dist/build                                                                                
dist/build/libHSparsers-0.12.5-BbWYl7dZaXs10a39DfSMvp-ghc8.2.1.so                         
dist/build/Text                                                                           
dist/build/Text/Parser                                                                    
dist/build/Text/Parser/Token.dyn_hi                                                       
dist/build/Text/Parser/Combinators.o                                                      
dist/build/Text/Parser/LookAhead.o                                                        
dist/build/Text/Parser/Char.dyn_o                                                         
dist/build/Text/Parser/Permutation.dyn_o                                                  
dist/build/Text/Parser/Expression.dyn_o                                                   
dist/build/Text/Parser/Expression.dyn_hi                                                  
dist/build/Text/Parser/Permutation.o                                                      
dist/build/Text/Parser/Char.dyn_hi                                                        
dist/build/Text/Parser/Char.hi
dist/build/Text/Parser/LookAhead.dyn_o
dist/build/Text/Parser/Expression.o
dist/build/Text/Parser/Combinators.dyn_o
dist/build/Text/Parser/Token.hi
dist/build/Text/Parser/Char.o
dist/build/Text/Parser/Expression.hi
dist/build/Text/Parser/Permutation.hi
dist/build/Text/Parser/Token.o
dist/build/Text/Parser/Token
dist/build/Text/Parser/Token/Highlight.dyn_o
dist/build/Text/Parser/Token/Highlight.o
dist/build/Text/Parser/Token/Highlight.hi
dist/build/Text/Parser/Token/Style.dyn_hi
dist/build/Text/Parser/Token/Style.dyn_o
dist/build/Text/Parser/Token/Highlight.dyn_hi
dist/build/Text/Parser/Token/Style.hi
dist/build/Text/Parser/Token/Style.o
dist/build/Text/Parser/LookAhead.hi
dist/build/Text/Parser/Combinators.dyn_hi
dist/build/Text/Parser/Combinators.hi
dist/build/Text/Parser/LookAhead.dyn_hi
dist/build/Text/Parser/Permutation.dyn_hi
dist/build/Text/Parser/Token.dyn_o
dist/build/libHSparsers-0.12.5-BbWYl7dZaXs10a39DfSMvp.a
dist/build/autogen
dist/build/autogen/cabal_macros.h
dist/build/autogen/Paths_parsers.hs
dist/build/doctests
dist/build/doctests/autogen
dist/build/doctests/autogen/cabal_macros.h
dist/build/doctests/autogen/Build_doctests.hs
dist/build/doctests/autogen/Paths_parsers.hs
dist/build/doctests/doctests-tmp
dist/build/doctests/doctests-tmp/Main.o
dist/build/doctests/doctests-tmp/Build_doctests.hi
dist/build/doctests/doctests-tmp/Main.hi
dist/build/doctests/doctests-tmp/Build_doctests.o
dist/build/doctests/doctests
dist/setup-config
dist/cabal-config-flags
dist/test
dist/test/parsers-0.12.5.log
dist/test/parsers-0.12.5-doctests.log
dist/setup
dist/setup/Main.o
dist/setup/setup.version
dist/setup/Main.hi
dist/setup/setup
dist/setup/setup.hs
dist/setup/setup_macros.h
dist/hpc
dist/hpc/vanilla
dist/hpc/vanilla/tix
dist/hpc/vanilla/tix/doctests
rgscott@gearloose:~/.../Hacking/Haskell/parsers-0.12.5$ find dist
dist
dist/package.conf.inplace
dist/package.conf.inplace/parsers-0.12.5-BbWYl7dZaXs10a39DfSMvp.conf
dist/package.conf.inplace/package.cache.lock
dist/package.conf.inplace/package.cache
dist/build
dist/build/libHSparsers-0.12.5-BbWYl7dZaXs10a39DfSMvp-ghc8.2.1.so
dist/build/Text
dist/build/Text/Parser
dist/build/Text/Parser/Token.dyn_hi
dist/build/Text/Parser/Combinators.o
dist/build/Text/Parser/LookAhead.o
dist/build/Text/Parser/Char.dyn_o
dist/build/Text/Parser/Permutation.dyn_o
dist/build/Text/Parser/Expression.dyn_o
dist/build/Text/Parser/Expression.dyn_hi
dist/build/Text/Parser/Permutation.o
dist/build/Text/Parser/Char.dyn_hi
dist/build/Text/Parser/Char.hi
dist/build/Text/Parser/LookAhead.dyn_o
dist/build/Text/Parser/Expression.o
dist/build/Text/Parser/Combinators.dyn_o
dist/build/Text/Parser/Token.hi
dist/build/Text/Parser/Char.o
dist/build/Text/Parser/Expression.hi
dist/build/Text/Parser/Permutation.hi
dist/build/Text/Parser/Token.o
dist/build/Text/Parser/Token
dist/build/Text/Parser/Token/Highlight.dyn_o
dist/build/Text/Parser/Token/Highlight.o
dist/build/Text/Parser/Token/Highlight.hi
dist/build/Text/Parser/Token/Style.dyn_hi
dist/build/Text/Parser/Token/Style.dyn_o
dist/build/Text/Parser/Token/Highlight.dyn_hi
dist/build/Text/Parser/Token/Style.hi
dist/build/Text/Parser/Token/Style.o
dist/build/Text/Parser/LookAhead.hi
dist/build/Text/Parser/Combinators.dyn_hi
dist/build/Text/Parser/Combinators.hi
dist/build/Text/Parser/LookAhead.dyn_hi
dist/build/Text/Parser/Permutation.dyn_hi
dist/build/Text/Parser/Token.dyn_o
dist/build/libHSparsers-0.12.5-BbWYl7dZaXs10a39DfSMvp.a
dist/build/autogen
dist/build/autogen/cabal_macros.h
dist/build/autogen/Paths_parsers.hs
dist/build/doctests
dist/build/doctests/autogen
dist/build/doctests/autogen/cabal_macros.h
dist/build/doctests/autogen/Build_doctests.hs
dist/build/doctests/autogen/Paths_parsers.hs
dist/build/doctests/doctests-tmp
dist/build/doctests/doctests-tmp/Main.o
dist/build/doctests/doctests-tmp/Build_doctests.hi
dist/build/doctests/doctests-tmp/Main.hi
dist/build/doctests/doctests-tmp/Build_doctests.o
dist/build/doctests/doctests
dist/setup-config
dist/cabal-config-flags
dist/test
dist/test/parsers-0.12.5.log
dist/test/parsers-0.12.5-doctests.log
dist/setup
dist/setup/Main.o
dist/setup/setup.version
dist/setup/Main.hi
dist/setup/setup
dist/setup/setup.hs
dist/setup/setup_macros.h
dist/hpc
dist/hpc/vanilla
dist/hpc/vanilla/tix
dist/hpc/vanilla/tix/doctests
```

--- a/README.md
+++ b/README.md
@@ -82,17 +82,19 @@ test-suite doctests:
 Notes
 -----
 
-* `Cabal-2.0` can be finicky about whether it builds the library component or
-  the `doctests` test suite component first. However, in order for
+* Recent versions of `Cabal` (for instance, 2.0) can choose to build a
+  package's `doctest` test suite _before_ the library. However, in order for
   `cabal-doctest` to work correctly, the library _must_ be built first, as
   `doctest` relies on the presence of generated files that are only created
   when the library is built. See
   [#19](https://github.com/phadej/cabal-doctest/issues/19).
 
-  To avoid headaches, it is recommended that you depend on the library itself
-  in the `doctests` test suite. See
+  A hacky workaround for this problem is to depend on the library itself in a
+  `doctests` test suite. See
   [the example's .cabal file](https://github.com/phadej/cabal-doctest/blob/master/example/example.cabal)
-  for a demonstration.
+  for a demonstration. (This assumes that the test suite has the ability to
+  read build artifacts from the library, a separate build component. In
+  practice, this assumption holds, which is why this library works at all.)
 
 * `custom-setup` section is supported starting from `cabal-install-1.24`.
   For older `cabal-install's` you have to install custom setup dependencies

--- a/README.md
+++ b/README.md
@@ -82,6 +82,18 @@ test-suite doctests:
 Notes
 -----
 
+* `Cabal-2.0` can be finicky about whether it builds the library component or
+  the `doctests` test suite component first. However, in order for
+  `cabal-doctest` to work correctly, the library _must_ be built first, as
+  `doctest` relies on the presence of generated files that are only created
+  when the library is built. See
+  [#19](https://github.com/phadej/cabal-doctest/issues/19).
+
+  To avoid headaches, it is recommended that you depend on the library itself
+  in the `doctests` test suite. See
+  [the example's .cabal file](https://github.com/phadej/cabal-doctest/blob/master/example/example.cabal)
+  for a demonstration.
+
 * `custom-setup` section is supported starting from `cabal-install-1.24`.
   For older `cabal-install's` you have to install custom setup dependencies
   manually.
@@ -93,11 +105,6 @@ Notes
 * There is [an issue in the Cabal issue tracker](https://github.com/haskell/cabal/issues/2327)
   about adding `cabal doctest` command. After that command is implemented,
   this library will be deprecated.
-
-* If your library contains `cbits`, you might need to depend on the library
-  itself in `doctests` test-suite. We aren't sure whether this a bug or not.
-  See [#5 issue](https://github.com/phadej/cabal-doctest/issues/5) for longer
-  explanation.
 
 * You can use `x-doctest-options` field in `test-suite doctests` to
   pass additional flags to the `doctest`.

--- a/example/example.cabal
+++ b/example/example.cabal
@@ -28,7 +28,8 @@ test-suite doctests
   main-is:             doctests.hs
   build-depends:
     base,
-    doctest >= 0.11.3 && <0.12,
+    doctest >= 0.11.3 && <0.13,
+    example,
     QuickCheck ==2.10.*,
     template-haskell
   ghc-options:         -Wall -threaded

